contracts/schemes/TokenTrade.sol

```diff
@@ -0,0 +1,179 @@
+    * @return bool success
+    */
+    function executeProposal(bytes32 _proposalId, int256 _decision)
+    external
+    onlyVotingMachine(_proposalId)
+    override
+    returns(bool) {
+        Proposal memory proposal = proposals[_proposalId];
+        if (_decision == 1) {
+            proposals[_proposalId].passed = true;
+        }
+        proposals[_proposalId].decided = true;
+
+        emit ProposalExecuted(address(avatar), _proposalId, _decision);
+        return true;
+    }
+
+
+    function execute(bytes32 _proposalId) public {
+        Proposal storage proposal = proposals[_proposalId];
+        require(proposal.exist, "must be a live proposal");
+        require(proposal.decided, "must be a decided proposal");
+        if (proposal.passed) {
+            proposal.exist = false;
+            proposal.sendToken.safeTransfer(address(avatar), proposal.sendTokenAmount);
+            require(
+                Controller(avatar.owner()).externalTokenTransfer(
+                    proposal.receiveToken, proposal.beneficiary, proposal.receiveTokenAmount
+                ), "failed to transfer tokens from the DAO"
+            );
+
+            emit TokenTradeProposalExecuted(
+                address(avatar),
+                _proposalId,
+                proposal.beneficiary,
+                proposal.sendToken,
+                proposal.sendTokenAmount,
+                proposal.receiveToken,
+                proposal.receiveTokenAmount
+            );
+            delete proposals[_proposalId];
+        } else {
+            Proposal memory _proposal = proposals[_proposalId];
+            delete proposals[_proposalId];
+            _proposal.sendToken.safeTransfer(address(_proposal.beneficiary), _proposal.sendTokenAmount);
+        }
+    }
+
+    /**
+    * @dev propose to trade tokens with the DAO
+    * @param _sendToken token the proposer suggests to send to the DAO
+    * @param _sendTokenAmount token amount the proposer suggests to send to the DAO
+    * @param _receiveToken token the proposer asks to receive from the DAO
+    * @param _receiveTokenAmount token amount the proposer asks to receive from the DAO
+    * @param _descriptionHash proposal description hash
+    * @return an id which represents the proposal
+    */
+    function proposeTokenTrade(
+        IERC20 _sendToken,
+        uint256 _sendTokenAmount,
+        IERC20 _receiveToken,
+        uint256 _receiveTokenAmount,
+        string memory _descriptionHash
+    )
+    public
+    returns(bytes32)
+    {
+        require(
+            address(_sendToken) != address(0) && address(_receiveToken) != address(0),
+            "Token address must not be null"
+        );
+        require(_sendTokenAmount > 0 && _receiveTokenAmount > 0, "Token amount must be greater than 0");
+
+        _sendToken.safeTransferFrom(msg.sender, address(this), _sendTokenAmount);
+        bytes32 proposalId = votingMachine.propose(2, voteParamsHash, msg.sender, address(avatar));
```

Comment:
proposalId can be defined as return value at https://github.com/daostack/arc/pull/774/files#diff-0568445d702e1cdbb17b4e3cdc934e5dR142